src/main/java/jenkins/scm/api/trait/SCMSourceBuilder.java

```diff
@@ -0,0 +1,200 @@
+ * <li>All methods should be either {@code final} or {@code abstract} unless there is a documented reason for
+ * allowing overrides</li>
+ * <li>All "setter" methods will return {@link B} and be called "withXxx"</li>
+ * <li>All "getter" methods will called "xxx()" callers should not assume that the returned value is resistent
+ * from concurrent changes but implementations should ensure that the returned value is immutable by the caller.
+ * In other words, it is intentional to reduce intermediate allocations by
+ * {@code return Collections.unmodifiableList(theList);} rather than the concurrency safe
+ * {@code return Collections.unmodifiableList(new ArrayList<>(theList));}
+ * </li>
+ * </ul>
+ *
+ * @param <B> the type of {@link SCMSourceBuilder} so that subclasses can chain correctly in their
+ *            {@link #withTrait(SCMSourceTrait)}  etc methods.
+ * @param <S> the type of {@link SCMSource} returned by {@link #build()}. In general this should be a type that
+ *            has {@link SCMSourceTrait} support.
+ */
+public abstract class SCMSourceBuilder<B extends SCMSourceBuilder<B, S>, S extends SCMSource> {
+
+    /**
+     * The base class of {@link SCMSource} that will be produced by the {@link SCMSourceBuilder}.
+     */
+    @NonNull
+    private final Class<S> clazz;
+    /**
+     * The project name.
+     */
+    @NonNull
+    private final String projectName;
+    /**
+     * The {@link SCMSourceTrait} instances to provide to the {@link SCMSource} (assuming the {@link SCMSource} is
+     * {@link SCMSourceTrait} aware.
+     */
+    @NonNull
+    private final List<SCMSourceTrait> traits = new ArrayList<SCMSourceTrait>();
+
+    /**
+     * Constructor.
+     * @param clazz the base class of {@link SCMSource} that will be produced by the {@link SCMSourceBuilder}.
+     * @param projectName the project name.
+     */
+    public SCMSourceBuilder(@NonNull Class<S> clazz, @NonNull String projectName) {
+        this.clazz = clazz;
+        this.projectName = projectName;
+    }
+
+    /**
+     * Returns the base class of {@link SCMSource} that will be produced by the {@link SCMSourceBuilder}.
+     * @return the base class of {@link SCMSource} that will be produced by the {@link SCMSourceBuilder}.
+     */
+    @NonNull
+    public final Class<S> sourceClass() {
+        return clazz;
+    }
+
+    /**
+     * Returns the project name.
+     * @return the project name.
+     */
+    @NonNull
+    public final String projectName() {
+        return projectName;
+    }
+
+    /**
+     * Returns the {@link SCMSourceTrait} instances to provide to the {@link SCMSource} (assuming the {@link SCMSource}
+     * is {@link SCMSourceTrait} aware.
+     * @return the {@link SCMSourceTrait} instances to provide to the {@link SCMSource} (assuming the
+     * {@link SCMSource} is {@link SCMSourceTrait} aware.
+     */
+    @NonNull
+    public final List<SCMSourceTrait> traits() {
+        return Collections.unmodifiableList(traits);
```

Comment:
https://github.com/jenkinsci/scm-api-plugin/pull/36/commits/e4003a8365e613cba88f6e3cd0e3d379847bf691 should address your concerns